src/vulkan-renderer/world/collision.cpp

```diff
@@ -0,0 +1,184 @@
+
+#include <glm/geometric.hpp>
+#include <glm/gtx/vector_angle.hpp>
+
+#include <array>
+#include <limits>
+#include <memory>
+#include <utility>
+
+namespace inexor::vulkan_renderer::world {
+
+template <typename T>
+RayCubeCollision<T>::RayCubeCollision(RayCubeCollision &&other) noexcept : m_cube{other.m_cube} {
+    m_intersection = other.m_intersection;
+    m_selected_face = std::move(other.m_selected_face);
+    m_nearest_corner = std::move(other.m_nearest_corner);
+    m_nearest_edge = std::move(other.m_nearest_edge);
+}
+
+template <typename T>
+RayCubeCollision<T>::RayCubeCollision(const T &cube, const glm::vec3 ray_pos, const glm::vec3 ray_dir) : m_cube(cube) {
+
+    // This lambda adjusts the center points on a cube's face to the size of the octree,
+    // so collision works with cubes of any size. This does not yet account for rotations!
+    const auto adjust_coordinates = [=](const glm::vec3 pos) {
+        // TODO: Take rotation of the cube into account.
+        return m_cube.center() + pos * (m_cube.size() / 2);
+    };
+
+    const std::array cube_directions{
+        glm::vec3(-1.0f, 0.0f, 0.0f), // left
+        glm::vec3(1.0f, 0.0f, 0.0f),  // right
+        glm::vec3(0.0f, -1.0f, 0.0f), // front
+        glm::vec3(0.0f, 1.0f, 0.0f),  // back
+        glm::vec3(0.0f, 0.0f, 1.0f),  // top
+        glm::vec3(0.0f, 0.0f, -1.0f)  // bottom
+    };
+
+    // The coordinates of the center of every face of the cube.
+    const std::array cube_face_centers{cube_face{"left", adjust_coordinates(cube_directions[0])},
+                                       cube_face{"right", adjust_coordinates(cube_directions[1])},
+                                       cube_face{"front", adjust_coordinates(cube_directions[2])},
+                                       cube_face{"back", adjust_coordinates(cube_directions[3])},
+                                       cube_face{"top", adjust_coordinates(cube_directions[4])},
+                                       cube_face{"bottom", adjust_coordinates(cube_directions[5])}};
+
+    // The coordinates of every corner of the cube.
+    std::array cube_corners = {cube_corner{"left back top", adjust_coordinates({-1.0f, 1.0f, 1.0f})},
+                               cube_corner{"right back top", adjust_coordinates({1.0f, 1.0f, 1.0f})},
+                               cube_corner{"left back botom", adjust_coordinates({-1.0f, 1.0f, -1.0f})},
+                               cube_corner{"right back bottom", adjust_coordinates({1.0f, 1.0f, -1.0f})},
+                               cube_corner{"left front top", adjust_coordinates({-1.0f, -1.0f, 1.0f})},
+                               cube_corner{"right front top", adjust_coordinates({1.0f, -1.0f, 1.0f})},
+                               cube_corner{"left front bottom", adjust_coordinates({-1.0f, -1.0f, -1.0f})},
+                               cube_corner{"right front bottom", adjust_coordinates({1.0f, -1.0f, -1.0f})}};
```

Comment:
Does not match with our [cube documentation](https://inexor-vulkan-renderer.readthedocs.io/en/latest/development/reference/octree-file-format.html). As a cube stays always positive in coordinates?

Reply:
The coordinates here are adjusted so the center of the cube is `(0,0,0)`. Otherwise we would have to do some anoying coordinate transformation for all the math functions we use.